static/app/utils/api/useAggregatedQueryKeys.tsx

```diff
@@ -0,0 +1,114 @@
+import {useCallback, useRef, useState} from 'react';
+
+import {ApiResult} from 'sentry/api';
+import {ApiQueryKey, fetchDataQuery, useQueryClient} from 'sentry/utils/queryClient';
+import useApi from 'sentry/utils/useApi';
+
+const BUFFER_WAIT_MS = 20;
+
+interface Props<QueryKeyAggregate, Data> {
+  /**
+   * Default data that goes into `useState`
+   */
+  defaultData: Data;
+
+  /**
+   * The queryKey reducer/generator.
+   *
+   * Takes the buffered "aggregates" and outputs an ApiQueryKey
+   */
+  genQueryKey: (ids: ReadonlyArray<QueryKeyAggregate>) => ApiQueryKey;
+
+  /**
+   * Data reducer, to integrate new requests with the previous state
+   */
+  reducer: (prev: Data, result: ApiResult<unknown>) => Data;
+
+  /**
+   * Callback should an error happen while fetching or reducing the data
+   */
+  onError?: (error: Error) => void;
+}
+
+export default function useAggregatedQueryKeys<QueryKeyAggregate, Data>({
```

Comment:
I'd appreciate a comment that explains how to use this in a simple case. Or maybe the behavior could be documented in a test? Just skimming the code here, it's seems a bit daunting.